Test-Json: use JsonSchema.Net instead of NJsonSchema

gregsdennis opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

<!-- Summarize your PR between here and the checklist. -->
Updates the `Test-Json` cmdlet to use [JsonSchema.Net](https://www.nuget.org/packages/JsonSchema.Net) instead of [NJsonSchema](https://www.nuget.org/packages/NJsonSchema) in order to:

- use System.Text.Json instead of Newtonsoft
- support the latest JSON Schema drafts (up to 2020-12)

Resolves 

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

Resolves #18009

NJsonSchema only supports up to draft 7, which at this point is several years old and two versions behind.  Additionally, it relies on Newtonsoft.Json.

This change is a merely drop-in replacement of the implementation that provides validation.

The cmdlet API doesn't change, however the cmdlet will no longer support draft 4 schemas.  Draft 4 is a decade old at this point and the JSON Schema team (myself included) is encouraging tooling to only support draft 6/7 (6 is a subset of 7) and later.  (In the future, support for these should be dropped as newer versions are released.)  I'm not sure if this is considered a breaking change for this repo.

To assist users in their migration away from draft 4, the [alterschema draft migration tool](https://github.com/sourcemeta/alterschema) has been created.

I expect the existing test coverage is adequate to cover this change.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

gregsdennis committed:
updated Test-Json cmdlet to use JsonSchema.Net instead of NJsonSchema

gregsdennis commented:
I'm not sure if this is a problem, but I notice that all of the tests are written with single quotes.  Newtonsoft is happy to parse these, but System.Text.Json will not.  I suspect this is the source the test failures, but I'm not sure.  I could use some help getting the tests to run locally.

iSazonov commented:
@gregsdennis Yes, it is main breaking change in System.Text.Json. 

@SteveL-MSFT Do you agree to begin migration to System.Text.Json in the 7.4 milestone?

gregsdennis committed:
change quoting of test-json tests

gregsdennis committed:
rebuild files.wxs

gregsdennis commented:
There seem to be a lot more changes to files.wxs that I expected.  Maybe this is normal?

gregsdennis commented:
![image](https://user-images.githubusercontent.com/2676804/188521952-b7ada7fb-4331-4a83-ba9b-1428ec338ba0.png)

"invalid spacing" - but what should it be?  It tells me its wrong but not how to fix it.

gregsdennis committed:
adjust brace spacing to match examples in other commands

gregsdennis committed:
remove space between closing curly brace and closing parenthesis

iSazonov commented:
@gregsdennis Please finish replacing ' with " in tests since you have started the process.


> Invalid spacing

Try reformat:
```c#
var validationResults = _jschema.Validate(parsedJson, new ValidationOptions
{
    OutputFormat = OutputFormat.Basic
});

> files.wxs

You need to build win package locally and the script will publish new files.wxs in temp folder.

iSazonov commented:
@gregsdennis Maybe #11397 help you a bit.

gregsdennis commented:
> Please finish replacing ' with " in tests since you have started the process.

I expect you don't mean _all_ the tests.  Even just in that file seems overkill.  It seems there's a preference to have `"` wherever possible, so I only changed what was necessary.

> Try reformat

The format of `Func(new Class { Prop = value });` seems to be working.  I can still change to what you have if that's preferred.

iSazonov commented:
> > Please finish replacing ' with " in tests since you have started the process.
> 
> I expect you don't mean _all_ the tests. Even just in that file seems overkill. It seems there's a preference to have `"` wherever possible, so I only changed what was necessary.

I still see test fails because of "

gregsdennis commented:
The failing tests aren't (all) because of `"`.

For example, `Json is valid against a valid schema from file` is failing because the lib doesn't know about `valid_schema_definitions.json`, which is being referenced.

JSON Schema explicitly states that implementations SHOULD NOT download references that appear to be URLs.  The URIs that are used are merely identifiers, not locations.  As such, downloading is disabled in my lib by default, requiring pre-loading of all required schemas, but it seems that it may have been enabled in NJsonSchema.  I can make this change.

I'll go through the other tests to see what the problems are.

iSazonov commented:
> As such, downloading is disabled in my lib by default, requiring pre-loading of all required schemas

So we cannot reference base schema file from new schema file?

iSazonov commented:
> If PowerShell is comfortable with automatically fetching external documents, it can be enabled, yes, but it's not generally recommended by JSON Schema. Given that this is previous functionality, I think enabling it is fine.
> 
> I suggest adding this to the docs as a deviation from the spec.

Since this has not been reported as a security issue until now we should keep current behavior if possible. Then we will ask PowerShell security group.

gregsdennis committed:
configure for automatic downloads of external schemas

gregsdennis committed:
use single quote strings when writing literal JSON values

gregsdennis commented:
It looks like there's some syntax issue I don't understand in the `Test-Json recognizes non-object types: <name>` tests on line 156.

I've extracted the loop into a separate file:

```powershell
$types = 'string', 'number', 'boolean', 'null', 'array', 'object'
$types | Where-Object {
    $schema = "{ ""type"": ""$_"" }"
    Write-Output $schema
}


$type = 'string'
$schema = "{ ""type"": ""$type"" }"
Write-Output $schema
```

This outputs 

```
string
number
boolean
null
array
object
{ "type": "string" }
```

I don't understand why the interpolation isn't working in the loop but it works outside of it.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Test-Json.Tests.ps1:
```diff
@@ -172,7 +172,7 @@ Describe "Test-Json" -Tags "CI" {
         $types = 'string', 'number', 'boolean', 'null', 'array', 'object'
         $types | Where-Object {
-            $schema = "{ 'type': '$_' }"
+            $schema = '{ "type": "$_" }'
```
```suggestion
            $schema = "{ `"type`": `"$_`" }"
```

iSazonov replied:
The test is not failed after my fix.

gregsdennis replied:
I'm confused.  If I could get an explanation on this outside of this PR, I'd be grateful.  In my extracted code it doesn't work unless I change `Where-Object` to `ForEach-Object`.

iSazonov committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Test-Json.Tests.ps1

iSazonov commented:
> I don't understand why the interpolation isn't working

PowerShell interpolation works only in double quotes.

I added commit to fix the test.

gregsdennis commented:
What's the policy on changing error codes?

`Test-Json throw if a schema from file is invalid` is expecting `InvalidJsonSchema`, but the file isn't valid JSON at all.  It seems that NJsonSchema preprocesses `$ref`s so that it tries to deserialize the text as a schema before it ever reaches the validation stage.  JsonSchema.Net deserializes referenced data (and caches it) at validation time.  This means that the command reaches the validation stage and the resulting error is `InvalidJson` instead (which I think is actually more correct).

In both cases, it's an error, but the exact error is different.

gregsdennis committed:
detect validation failures and report errors better

iSazonov commented:
> This means that the command reaches the validation stage and the resulting error is `InvalidJson` instead (which I think is actually more correct).

I agree with `InvalidJson` since we check just Json. However, the user must understand the root of the error whether it is the Json itself or a problem with the schema (reading/parsing). In other words `InvalidJson` must contain `InvalidJsonSchema` as an inner exception.

iSazonov commented on assets/wix/files.wxs:
```diff
@@ -3153,6 +3147,57 @@
         <File Id="fil48314E2AC49B4E208C5667CEC50CA78C" KeyPath="yes" Source="$(var.ProductSourcePath)\RegisterMicrosoftUpdate.ps1" />
       </Component>
+      <Component Id="cmp8AAE18271AD64A009017665757ED915B">
+        <File Id="filBD4DDF972FCA4F898D0ACABC42936DE6" KeyPath="yes" Source="$(var.ProductSourcePath)\Microsoft.PowerShell.SDK.pdb" />
```
We haven't .pdb files in msi before so please remove all from the file.

gregsdennis replied:
👍 

This is just what was built locally following the instructions for the _.wxs_ file.  Happy to remove things that aren't necessary.

gregsdennis replied:
Still getting [build failure](https://dev.azure.com/powershell/PowerShell/_build/results?buildId=108371&view=logs&j=ec161491-f990-5215-209c-5fb381a7fcfb&t=fd903d37-c62b-5a02-a186-3af73ce70275&l=2258) on this file.

```
Current files to not match  {D:\a\1\PowerShell\assets\wix\Files.wxs}
```

Not a very descriptive error.  It details where in the build script it failed, but not what the mismatch was.

iSazonov commented on assets/wix/files.wxs:
```diff
@@ -3153,6 +3147,57 @@
+        <File Id="filF4983633E7914A1AAD5D02BFDFA7DD63" KeyPath="yes" Source="$(var.ProductSourcePath)\JsonSchema.Net.dll" />
+      </Component>
+      <Component Id="cmp61D33FB49B2E4C7D8E9C5EFC3F3A04B2">
+        <File Id="fil78E3FEFE7B0E4FE1993AC9C117B582C5" KeyPath="yes" Source="$(var.ProductSourcePath)\JetBrains.Annotations.dll" />
```
I wonder we have to include the dependency.
I think that's the problem. I don't think that's what we want to distribute.

And what is official license for the file?

gregsdennis replied:
https://www.nuget.org/packages/JetBrains.Annotations/

License is MIT.

I've had requests to remove this before.  It's actually a dependency.  I use `[RegexParam]` in a few of my methods where a Regex string is expected.  It provides Regex syntax highlighting in the string literal.  If there's a way to get this without the runtime dependency, I'm unaware of it.

iSazonov replied:
> It provides Regex syntax highlighting in the string literal.

You could use .Net attribute https://source.dot.net/#System.Text.RegularExpressions/System/Text/RegularExpressions/Regex.cs,59

gregsdennis replied:
This will require an update to the library, so I'll have to update the version ref in the _.csproj_ file after I publish.  Shouldn't take long.

gregsdennis replied:
That's only available in .Net 7 preview: https://docs.microsoft.com/en-us/dotNet/API/system.diagnostics.codeanalysis.stringsyntaxattribute?view=net-7.0&viewFallbackFrom=netframework-4.8)  My library is .Net Standard 2.0.

iSazonov replied:
From [docs](https://www.jetbrains.com/help/resharper/Code_Analysis__Annotations_in_Source_Code.html):
>By default, all annotation attribute classes are marked with [Conditional("JETBRAINS_ANNOTATIONS")] attribute so the compiler will ignore the attribute usages in your code, which means that no binary reference to the 'JetBrains.Annotations.dll' assembly will be produced. However, you can define 'JETBRAINS_ANNOTATIONS' conditional compilation symbol in your projects to preserve the attributes in the metadata.

So you can remove the dependency from binary package.

gregsdennis replied:
That removes the attribute from the compilation. That doesn't help my users, which is the entire point of including the attribute.  I don't need it for my own development; it's so that the people who use the library get the benefit.

iSazonov replied:
You can use [embed attribute declarations](https://www.jetbrains.com/help/resharper/Code_Analysis__Annotations_in_Source_Code.html#embedding-declarations-of-code-annotations-in-your-source-code)

iSazonov commented on assets/wix/files.wxs:
```diff
@@ -3153,6 +3147,57 @@
+        <File Id="fil3EA283F5D6624A879E39A4D771FF6B8F" KeyPath="yes" Source="$(var.ProductSourcePath)\Microsoft.WSMan.Management.pdb" />
+      </Component>
+      <Component Id="cmpB5F59B2105D54C0D896C8D7EB3EA82F3">
+        <File Id="filB6F08AEE602D4982A38EB3532F62BCF3" KeyPath="yes" Source="$(var.ProductSourcePath)\JsonPointer.Net.dll" />
```
Why don't you put all in one JsonSchema.Net.dll?

gregsdennis replied:
Because JsonPointer.Net is used by other libraries in my `json-everything` suite: https://github.com/gregsdennis/json-everything

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -140,23 +164,36 @@ protected override void ProcessRecord()
+
+                        if (validationResults.HasNestedResults)
+                        {
+                            foreach (var nestedResult in validationResults.NestedResults.Where(x => x.Message != null))
```
We prefer to avoid LINQ. Please replace with `if` operator in the loop body.

gregsdennis replied:
Done.

Curiousity: why avoid it?

iSazonov replied:
LINQ is too expensive sometimes.

gregsdennis commented:
> In other words InvalidJson must contain InvalidJsonSchema as an inner exception and the test should check this.

I don't know that this can be identified except by analyzing error messages.

I'll need to run some tests on my implementation to see what it's doing.  If it throws a `JsonException` during validation, it might be safe to assume that it's the result of attempting to deserialize a schema, and we can add that as a `catch` to the cmdlet.  Otherwise, I'm not sure we can capture this.

iSazonov commented:
Do you mean your JsonSchema.FromText() doesn't deserialize?

gregsdennis commented:
> Do you mean your JsonSchema.FromText() doesn't deserialize in call time?

It does.  I just needed to check what the behavior was.  I didn't have that test.  I've added it and `JsonException` is thrown as expected.  Will be adding a commit shortly.

gregsdennis commented:
> I agree with InvalidJson since we check just Json. However, the user must understand the root of the error whether it is the Json itself or a problem with the schema (reading/parsing). In other words InvalidJson must contain InvalidJsonSchema as an inner exception and the test should check this.

Do we want to use this proposed behavior or maintain the current?  Since it's throwing `JsonException`, I can isolate this case and make it do either per your preference.

iSazonov commented:
> Do we want to use this proposed behavior or maintain the current?

I think we should keep current behavior and have JsonSchemaException.

gregsdennis committed:
remove pdb files from files.wxs

gregsdennis committed:
capture deserialization exception and record as failure to read ref'd schema

iSazonov commented on assets/wix/files.wxs:
```diff
@@ -4166,6 +4170,23 @@
       <ComponentRef Id="cmpF3EB02BB62A34F509B22CC370C430A20" />
       <ComponentRef Id="cmp16FC1857F01A4E388616BA7AA8CC5F73" />
       <ComponentRef Id="cmp42A739F2AAC24D729DDBCAA096B8C55A" />
+      <ComponentRef Id="cmp8AAE18271AD64A009017665757ED915B" />
+      <ComponentRef Id="cmpB5AD8BDC2504469FADEFE3E6AA80BC27" />
+      <ComponentRef Id="cmpCCC817E4F2A04A73B4101A69417423A9" />
+      <ComponentRef Id="cmp8FDBF26F5E6F4BEB86E0FE5ED91CA638" />
+      <ComponentRef Id="cmp074B6F8EA20548DCA2FA446D9E27D3FE" />
+      <ComponentRef Id="cmp41A0A813CB2C476389F59D10C2E7D5B5" />
+      <ComponentRef Id="cmpE3D0DEFE9F454D7DA5296449DD21D484" />
+      <ComponentRef Id="cmpB7F854623ED34DB5B1B2C38D443BE731" />
+      <ComponentRef Id="cmp5B1AAAB6B46D45EA8A430A766325B0D5" />
+      <ComponentRef Id="cmp61D33FB49B2E4C7D8E9C5EFC3F3A04B2" />
+      <ComponentRef Id="cmpB51390A44A2E4D2EB27C891E4ED3A705" />
+      <ComponentRef Id="cmp2B4FE7A639B749F3B2814F87794CC3D5" />
+      <ComponentRef Id="cmpB5F59B2105D54C0D896C8D7EB3EA82F3" />
+      <ComponentRef Id="cmp611C894288294AC1A7D34FB420A338C6" />
+      <ComponentRef Id="cmp50CA3E6E9D0848CA86DF397DD6B4C0D6" />
+      <ComponentRef Id="cmp9D8D228F1E9A46899F26B3261EAF2341" />
+      <ComponentRef Id="cmp8467336B8279428087236432B11906AE" />
```
Please clean the list too (pdb!).

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -85,7 +109,7 @@ protected override void BeginProcessing()
-                        _jschema = JsonSchema.FromJsonAsync(Schema).Result;
+                        _jschema = JsonSchema.FromText(Schema);
                     }
                     catch (AggregateException ae)
```
Now it is not async and no AggregateException is. 

I'd expect your API throw there is a problem with the schema loading/parsing.

gregsdennis replied:
Yes, it'd be the JsonException mentioned before. 👍

gregsdennis replied:
Interestingly, the `catch` on [line 127](https://github.com/PowerShell/PowerShell/pull/18023/files#diff-75e882f1df5a078731c96023ef526ca48672b2d6d854008d1cc9d2182e44f990L127) (catching an invalid schema, e.g. a boolean in `maximum`) would also throw a `JsonException` because the data is validated during deserialization.  I don't think this `catch` would ever be hit.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -99,7 +123,7 @@ protected override void BeginProcessing()
-                        _jschema = JsonSchema.FromFileAsync(resolvedpath).Result;
+                        _jschema = JsonSchema.FromFile(resolvedpath);
                     }
                     catch (AggregateException ae)
```
The same.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -140,24 +164,51 @@ protected override void ProcessRecord()
+                            }
                         }
                     }
+                    catch (JsonException jsonExc)
+                    {
+                        result = false;
+
+                        Exception exception = new(TestJsonCmdletStrings.InvalidJsonSchema, jsonExc);
+                        WriteError(new ErrorRecord(exception, "InvalidJsonSchema", ErrorCategory.InvalidData, Json));
+                    }
```
I wonder if Validate() throw on schema but not JsonSchema.FromFile() or JsonSchema.FromText().

gregsdennis replied:
It would, but isolating the `.Validate()` is the only way to detect that it was a _referenced_ schema that failed to parse, so we still need this here.

iSazonov replied:
> it was a _referenced_ schema that failed to parse

Why isn't it parsed/validated at JsonSchema.FromText/FromFile time? I'd expect validate API never throw but return a result.
It may make sense to postpone this operation, but your API becomes somewhat difficult to understand, at least it violates expectations. In this context, it would make more sense if JsonSchemaException throws instead of JsonException.

gregsdennis replied:
There's some confusion.

`.FromText()`/`.FromFile()` load the initial schema by parsing & deserializing a single JSON text.  A failure here would raise a `JsonException`.  This is happening in `BeginProcessing()`.

If that schema has references (`$ref` and `$dynamicRef`) that lead to external files/locations, the references are resolved at validation time, which means that they must be deserialized at validation time, but that also yields a `JsonException`.  To distinguish between the instance (the JSON data being validated) being bad JSON and trying to deserialize a bad schema, we have to isolate the `.Validate()` method.  If `.Validate()` throws a `JsonException`, we know it must be from an attempt to deserialize a referenced schema.

iSazonov replied:
I think we should add such comment in code.

gregsdennis committed:
remove component ids for pdb files from .wxs file

gregsdennis commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -85,25 +88,23 @@ protected override void BeginProcessing()
-                        // Even if only one exception is thrown, it is still wrapped in an AggregateException exception
-                        // https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/exception-handling-task-parallel-library
-                        ae.Handle(UnwrapException);
+                        ExceptionDispatchInfo.Capture(e).Throw();
```
This was the applicable line of code that is in `UnwrapException`.  You can see that method has been removed.  I simply replaced the call with this line from that method.

iSazonov replied:
It was for unwrapping AggregateException. Now we haven't AggregateEception. Please WriteError().

gregsdennis committed:
remove unused usings; add comment regarding json exception during validation

gregsdennis commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -52,31 +55,31 @@ public class TestJsonCommand : PSCmdlet
 
-            return true;
-        }
+                return JsonSerializer.Deserialize<JsonSchema>(text);
```
Yes.  This is called while attempting to resolve a `$ref` or `$dynamicRef`, during validation.

iSazonov replied:
So we should catch all exceptions from HttpClient and File.ReadAllText().

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -140,24 +139,54 @@ protected override void ProcessRecord()
+                            }
                         }
                     }
+                    // A JsonException that occurs during validation is the result of a failure to deserialize
+                    // a referenced schema (through $ref or $dynamicRef).  References are resolved at validation
+                    // time and deserialized on-demand.
+                    catch (JsonException jsonExc)
```
Please move the comment in catch block to avoid style warning.

Also we should catch all exceptions here as commented above.
"So we should catch all exceptions from HttpClient and File.ReadAllText()."

gregsdennis replied:
There are a _lot_ of potential exceptions that come from those.  I just wrapped the fetch code in a try/catch and had it wrap the exception in a new one.  Then I could just catch that exception below.

gregsdennis committed:
add schema resolution exception; wrap fetch exceptions in new exception and handle

gregsdennis committed:
use WriteError instead of throwing exception

gregsdennis committed:
fix xml comments

iSazonov reviewed (commented):
@gregsdennis Please add more tests we don't cover all catch blocks.

gregsdennis commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/JsonSchemaReferenceResolutionException.cs:
```diff
@@ -0,0 +1,25 @@
+    /// (<code>Nothing</code> in Visual Basic) if no inner exception is specified.
+    /// </param>
+    public JsonSchemaReferenceResolutionException(Exception innerException)
+        : base("An error occurred attempting to resolve a schema reference", innerException)
```
I'm happy to move it to resx, but those were generate, and I haven't looked into how to do that.

iSazonov replied:
ErrorRecord contains already localized string so here null is ok.

gregsdennis commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -52,31 +53,38 @@ public class TestJsonCommand : PSCmdlet
-                ExceptionDispatchInfo.Capture(e.InnerException).Throw();
-            }
-            else
+            SchemaRegistry.Global.Fetch = uri =>
```
That's a thing?  Is that new?

gregsdennis replied:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/static-anonymous-functions 😮

iSazonov replied:
It is style. It says the delegate doesn't depend on the instance properties.

gregsdennis commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -140,26 +148,50 @@ protected override void ProcessRecord()
             try
             {
-                var parsedJson = JToken.Parse(Json);
+                var parsedJson = JsonNode.Parse(Json);
```
We can, but it'll be casted to a node anyway at the `Validate()` call.  It's better to let the code show what's happening.

iSazonov replied:
I don't remember why I used JsonDocument in my PR.

gregsdennis commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -140,26 +148,50 @@ protected override void ProcessRecord()
-                            WriteError(errorRecord);
+                            Exception exception = new(TestJsonCmdletStrings.InvalidJsonAgainstSchema);
+
+                            if (validationResults.Message != null)
+                            {
+                                ErrorRecord errorRecord = new(exception, "InvalidJsonAgainstSchema",
+                                    ErrorCategory.InvalidData, null);
```
I'm not sure how the newline was inserted, but I prefer to revert it to the previous single line.

gregsdennis commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -140,26 +148,50 @@ protected override void ProcessRecord()
+                            {
+                                foreach (var nestedResult in validationResults.NestedResults)
+                                {
+                                    if (nestedResult.Message == null)
+                                    {
+                                        continue;
+                                    }
+
+                                    ErrorRecord errorRecord = new(exception, "InvalidJsonAgainstSchema", ErrorCategory.InvalidData, null);
```
Why split this one?  The existing call is on a single line.

gregsdennis committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/JsonSchemaReferenceResolutionException.cs

Co-authored-by: Ilya <darpa@yandex.ru>

gregsdennis committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs

Co-authored-by: Ilya <darpa@yandex.ru>

gregsdennis committed:
use single-line method call; remove unused usings

gregsdennis commented:
@iSazonov are you looking for more PowerShell tests or a new set of c# tests?

iSazonov commented:
> @iSazonov are you looking for more PowerShell tests or a new set of c# tests?
> 
> (I can run the c# tests locally, but I haven't figured out running the PowerShell ones.)

I mean PowerShell tests. (We have C# tests only for some SDK API).

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/JsonSchemaReferenceResolutionException.cs:
```diff
@@ -0,0 +1,25 @@
+/// Thrown during evaluation of <see cref="TestJsonCommand"/> when an attempt
+/// to resolve a <code>$ref</code> or <code>$dynamicRef</code> fails.
+/// </summary>
+internal class JsonSchemaReferenceResolutionException : Exception
```
I guess we have to make this public to use in tests.

gregsdennis commented:
@iSazonov I'm looking through all of the catch blocks and the tests that exist, and I can't find any catches that aren't covered.  Which ones do you see?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:
```diff
@@ -140,26 +147,49 @@ protected override void ProcessRecord()
-                    }
                 }
             }
+            catch (JsonSchemaReferenceResolutionException jsonExc)
+            {
+                result = false;
+
+                Exception exception = new(TestJsonCmdletStrings.InvalidJsonSchema, jsonExc);
+                WriteError(new ErrorRecord(exception, "InvalidJsonSchema", ErrorCategory.InvalidData, _jschema));
+            }
```
Have we a test for the branch?

gregsdennis replied:
All of the cases where a referenced schema fails to load (and there are several) will flow through this `catch`.

iSazonov replied:
I don't see tests where we looks JsonSchemaReferenceResolutionException

gregsdennis replied:
We don't need to. That's an internal implementation detail. It's presented to the user as "invalid JSON schema."

iSazonov replied:
We throw InvalidJsonSchema in 3 places - for file, for text and if load from a reference - we haven't a test for the last. So we could create a string with valid schema but with bad reference (to non-existent file) and in InvalidJsonSchema exception check inner exception name for JsonSchemaReferenceResolutionException.

gregsdennis replied:
That would just be adding an assertion to an existing test. I can do that, but I don't think that it warrants new tests.

gregsdennis replied:
Do you have an example where inner exceptions are being checked in powershell?  I only see assertions on the error output.

```ps
$errorVar[0].FullyQualifiedErrorId | Should -BeExactly "InvalidJsonAgainstSchema,Microsoft.PowerShell.Commands.TestJsonCommand"
```

This doesn't mention the exceptions that were thrown.

gregsdennis replied:
> We throw InvalidJsonSchema in 3 places - for file, for text and if load from a reference - we haven't a test for the last.

`Test-Json throw if a schema from file is invalid` which throws an error because _invalid_schema_definitions.json_ is invalid JSON is failing to load a reference.  Granted, it's failing to resolve the reference because the referenced file is invalid JSON, but the reference resolution still fails, triggering a `JsonSchemaReferenceResolutionException`, which is caught on line 186 and repackaged as an "InvalidJsonSchema" error for the user.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/JsonSchemaReferenceResolutionException.cs:
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
```
```suggestion
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

#nullable enable

using System;
```

iSazonov commented:
@gregsdennis We moved to .Net 7 RC1. Please resolve merge conflicts.

gregsdennis committed:
enable nullable refs in JsonSchemaReferenceResolutionException.cs

Co-authored-by: Ilya <darpa@yandex.ru>

gregsdennis commented:
I've rebased and pushed.

gregsdennis commented:
@iSazonov since the update to .Net 7 RC1, I can't build anymore.  I've installed the [preview framework](https://dotnet.microsoft.com/en-us/download/dotnet/thank-you/sdk-7.0.100-rc.1-windows-x64-installer) (Windows x64 installer).

I'm getting a lot of errors saying that it can't find Microsoft.CodeAnalysis.

```
An instance of analyzer Microsoft.CodeAnalysis.MakeFieldReadonly.MakeFieldReadonlyDiagnosticAnalyzer
cannot be created from C:\Program Files\dotnet\sdk\7.0.100-rc.1.22431.12\Sdks\Microsoft.NET.Sdk\codestyle\cs\Microsoft.CodeAnalysis.CodeStyle.dll :
Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.4.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'
or one of its dependencies. The system cannot find the file specified.
```

It seems there are some rebase conflicts with how I merged the .wxs file.

iSazonov commented:
> I'm getting a lot of errors saying that it can't find Microsoft.CodeAnalysis.

Usually we need to remove nuget cache in current user profile to fix such issues. Also you need to run once `Start-PSBuild -Clean`.

gregsdennis committed:
add removed files from wxs

gregsdennis committed:
add more removed files from wxs

gregsdennis committed:
remove njsonschema from wxs

gregsdennis committed:
regenerate files.wxs

gregsdennis commented:
@iSazonov I've regen'd the files.wxs file (and removed the pdb's) but I'm not sure it's right.  There seem to be other DLLs listed that I didn't do anything with.

iSazonov commented:
> @iSazonov I've regen'd the files.wxs file (and removed the pdb's) but I'm not sure it's right. There seem to be other DLLs listed that I didn't do anything with.

Please build msi package locally and use files.wxs from the build.

gregsdennis commented:
> Please build msi package locally and use files.wxs from the build.

Are there instructions for this?  All I can find is the [Windows build instructions](https://github.com/PowerShell/PowerShell/blob/master/docs/building/windows-core.md).  This only generates an _.exe_.

iSazonov commented:
https://github.com/PowerShell/PowerShell/blob/7dc4587014bfa22919c933607bf564f0ba53db2e/.github/CONTRIBUTING.md#pull-request---automatic-checks
```powershell
Import-Module .\build.psm1
Start-PSBuild -Clean -CrossGen -PSModuleRestore -Runtime win7-x64 -Configuration Release -ReleaseTag <release tag>
Import-Module .\tools\packaging
Start-PSPackage -Type msi -ReleaseTag <release tag> -WindowsRuntime 'win7-x64' -SkipReleaseChecks
```
Console output should contain a reference to new files.wxs.

gregsdennis committed:
set files.wxs to match local build

gregsdennis commented:
@iSazonov I ran that except that it didn't recognize the `-CrossGen` switch, so I had to remove that.

```
PS C:\projects\PowerShell> Start-PSBuild -Clean -CrossGen -PSModuleRestore -Runtime win7-x64 -Configuration Release -ReleaseTag v1.0.0-local
Start-PSBuild: A parameter cannot be found that matches parameter name 'CrossGen'.
```

It also added all of the `.pdb` files, so I removed those.  It still doesn't match. 😭

iSazonov committed:
Fix files.wxs

iSazonov commented:
@gregsdennis I added new commit to fix packaging.

gregsdennis commented:
Thank you @iSazonov 

I'm going through the checklist now, and the only breaking change that I can think of is the lack of support for single-quoted JSON (which isn't really JSON anyway).

Does this need to be documented somewhere?  If so, where?  Do I need to create a separate issue to track that?

iSazonov commented:
> Does this need to be documented somewhere? If so, where? Do I need to create a separate issue to track that?

Common rule for PowerShell is to follow .Net behavior. I think it makes sense to explain the breaking change in docs. But now is not the time. The change has to be approved by MSFT team first. Then there are other Json cmdlets that should be migrated. And finally, there are many small differences between NewtonSoft and .Net implementations. I guess we are not able to describe them all. We can only point out this fact. I don't think this is a critical change, because the .Net implementations follow Json standard very closely.

@gregsdennis The PR contains a lot of temporary comments, so I suggest to close the PR and open new one.
